modules/modelSetup/mixin/ModelSetupDiffusionLossMixin.py

```diff
@@ -149,28 +150,75 @@ def __unmasked_losses(
             losses /= mask_mean
 
         return losses
-
-    def __min_snr_weight(
+    
+    def calculate_loss(
             self,
-            timesteps: Tensor,
-            gamma: float,
-            v_prediction: bool,
-            device: torch.device
-    ):
+            model,
+            batch: dict,
+            data: dict,
+            config: TrainConfig,
+    ) -> Tensor:
+        return self._diffusion_losses(
+            batch=batch,
+            data=data,
+            config=config,
+            train_device=self.train_device,
+            betas=model.noise_scheduler.betas.to(device=self.train_device),
+        ).mean()
+    
+    def __snr(self, timesteps: Tensor, device: torch.device):
         if self.__coefficients:
             all_snr = (self.__coefficients.sqrt_alphas_cumprod /
                        self.__coefficients.sqrt_one_minus_alphas_cumprod) ** 2
             all_snr.to(device)
             snr = all_snr[timesteps]
         else:
             alphas_cumprod = self.__alphas_cumprod_fun(timesteps, 1)
-            snr = ((alphas_cumprod ** 0.5) / (1.0 - alphas_cumprod) ** 0.5) ** 2
+            snr = alphas_cumprod / (1.0 - alphas_cumprod)
+        
+        return snr
+        
+
+    def __min_snr_weight(
+            self,
+            timesteps: Tensor,
+            gamma: float,
+            v_prediction: bool,
+            device: torch.device
+    ) -> Tensor:
+        snr = self.__snr(timesteps, device)
         min_snr_gamma = torch.minimum(snr, torch.full_like(snr, gamma))
         # Denominator of the snr_weight increased by 1 if v-prediction is being used.
         if v_prediction:
-            snr += 1
-        snr_weight = torch.div(min_snr_gamma, snr).float().to(device)
+            snr += 1.0
+        snr_weight = (min_snr_gamma / snr).to(device)
         return snr_weight
+    
+    def __debiased_estimation_weight(
+        self,
+        timesteps: Tensor,
+        v_prediction: bool,
+        device: torch.device
+    ) -> Tensor:
+        snr = self.__snr(timesteps, device)
+        weight = snr
+        torch.clip(weight, min=1.0e-3, max=1.0e3, out=weight)
```

Comment:
I don't think you should clip at 1e-3 anymore. SNR+1 covers the zero-terminal case more elegantly. Correct me if I'm wrong but I don't recall them doing that clipping in the paper. For that matter, I assume the 1e3 clipping is for compatibility with Kohya's implementation rather than done as per the paper?

Reply:
Done in https://github.com/Nerogar/OneTrainer/pull/212/commits/46e7d155bfda23bd6fb20f815bad31ff34816bc4.